evm/p2p/p2p_proto.py

```diff
@@ -56,10 +56,10 @@ def get_reason_name(self, reason_id):
 
     def handle(self, proto, data):
         decoded = self.decode(data)
-        reason_name = self.get_reason_name(decoded['reason'])
+        decoded['reason_name'] = self.get_reason_name(decoded['reason'])
```

Comment:
what do you think about just pulling the name out for the logging message, but not mutating the `decoded` value. 

Alternate solution would be to use `cytools.assoc` to do this in a non-mutating way

```
raw_decoded = self.decode(data)
decoded = assoc(raw_decoded, 'reason_name', self.get_reason_name(raw_decoded['reason']))
```

Reply:
I include it in the `decoded` dict because it is now used in `handshake()` so that it can include the disconnect reason in the exception it raises.

I don't understand what's the benefit of using cytools.assoc here?

Reply:
> I don't understand what's the benefit of using cytools.assoc here?

It's more about consistency in avoiding mutation.  In this case there is no direct benefit, but by sticking to the pattern it keeps the code consistent. Largely an ideology based recommendation.

An alternative approach would be to override `decode` on this object and add the inclusion of the `reason_name` there so that it's part of the decoding (which feels slightly more correct).

Reply:
I'm fine with using cytooz.assoc, just wanted to understand the reasoning behind it. If we were to override `decode()` here you'd still prefer if we used assoc() there though?